presto-orc/src/main/java/com/facebook/presto/orc/writer/FlatMapColumnWriter.java

```diff
@@ -0,0 +1,289 @@
+    private final int column;
+    private final int sequence;
+    private final int keyColumnId;
+    private final int valueColumnId;
+    /* Type data for for the Map */
+    private final MapType type;
+    private final Type keyType;
+    private final Type valueType;
+    /* We need the OrcType of the key to perform Type specific Block writes */
+    private final OrcType keyOrcType;
+    private final boolean compressed;
+    /* Column Encoding data for the Map */
+    private final ColumnEncoding columnEncoding;
+    private final ColumnEncoding keyColumnEncoding;
+    private final ColumnEncoding valueColumnEncoding;
+    private final Optional<DwrfDataEncryptor> dwrfEncryptor;
+    /* Metadata writer for the Map type and copy of the passed metadata writer to pass to FlatMapKeyNode writer */
+    private final CompressedMetadataWriter metadataWriter;
+    private final MetadataWriter inputMetadataWriter;
+    /* valueWriterFactory to generate streams for individual keys */
+    private final ColumnWriterFactory valueWriterFactory;
+    private final int stringStatisticsLimitInBytes;
+
+    private final LongOutputStream lengthStream;
+    private final PresentOutputStream presentStream;
+    private List<FlatMapKeyNodeWriter> keyNodes;
+    /* Block Writer for Numeric or String Keyed Types */
+    private final FlatMapBlockWriter blockWriter;
+    private StatisticsBuilder statisticsBuilder;
+    private final Supplier<StatisticsBuilder> statisticsBuilderSupplier;
+
+    private final List<ColumnStatistics> rowGroupColumnStatistics = new ArrayList<>();
+    private long columnStatisticsRetainedSizeInBytes;
+
+    private int nonNullValueCount;
+    private boolean closed;
+
+    private StringStatisticsBuilder newStringStatisticsBuilder()
+    {
+        return new StringStatisticsBuilder(stringStatisticsLimitInBytes);
+    }
+
+    public FlatMapColumnWriter(
+            int column,
+            int sequence,
+            ColumnWriterOptions columnWriterOptions,
+            Optional<DwrfDataEncryptor> dwrfEncryptor,
+            OrcEncoding orcEncoding,
+            MetadataWriter metadataWriter,
+            Type type,
+            List<OrcType> orcTypes,
+            ColumnWriterFactory valueWriterfactory)
+    {
+        checkArgument(column >= 0, "column is negative");
+        checkArgument(sequence >= 0, "sequence is negative");
+
+        this.column = column;
+        this.sequence = sequence;
+        this.keyColumnId = orcTypes.get(column).getFieldTypeIndex(0);
+        this.valueColumnId = orcTypes.get(column).getFieldTypeIndex(1);
+
+        this.type = (MapType) type;
+        this.keyType = type.getTypeParameters().get(0);
+        this.valueType = type.getTypeParameters().get(1);
+
+        this.keyOrcType = orcTypes.get(keyColumnId);
+
+        this.compressed = columnWriterOptions.getCompressionKind() != NONE;
+
+        this.columnEncoding = new ColumnEncoding(orcEncoding == DWRF ? DWRF_MAP_FLAT : DIRECT_V2, 0);
```

Comment:
Flat Map is of DWRF type, so all these conditions should be optimized.

Add a check the input orCnecoding is of DWRF type.